File: packages/sdk-js/src/context.ts
/**
 * XASE SDK - Context Capture
 * 
 * Captures runtime context for evidence records
 */

import { createHash } from 'crypto'
import { hostname } from 'os'

/**
 * Captures current runtime context
 */
export function captureContext(): Record<string, any> {
  return {
    runtime: `node@${process.version}`,
    platform: process.platform,
    arch: process.arch,
    hostname: hostname(),
    pid: process.pid,
    libVersion: '0.1.0', // TODO: Read from package.json
    env: process.env.NODE_ENV || 'development',
    timestamp: Date.now(),
  }
}

/**
 * Generates a stable idempotency key from payload
 */
export function generateIdempotencyKey(data: string): string {
  const hash = createHash('sha256')
  hash.update(data)
  return hash.digest('hex').substring(0, 32)
}

/**
 * Validates idempotency key format
 */
export function isValidIdempotencyKey(key: string): boolean {
  // UUID v4
  const uuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i
  if (uuidRegex.test(key)) {
    return true
  }

  // Alphanumeric 16-64 chars
  const alphanumericRegex = /^[a-zA-Z0-9_-]{16,64}$/
  return alphanumericRegex.test(key)
}

File: packages/sdk-py/src/xase/context.py
"""
XASE SDK - Context Capture

Captures runtime context for evidence records.
"""

import hashlib
import os
import platform
import re
import socket
import sys
from typing import Any, Dict


def capture_context() -> Dict[str, Any]:
    """Capture current runtime context."""
    return {
        "runtime": f"python@{sys.version.split()[0]}",
        "platform": platform.system(),
        "arch": platform.machine(),
        "hostname": socket.gethostname(),
        "pid": os.getpid(),
        "lib_version": "0.1.0",
        "env": os.getenv("ENV") or os.getenv("PYTHON_ENV") or "development",
        "timestamp": int(__import__("time").time() * 1000),
    }


def generate_idempotency_key(data: str) -> str:
    """Generate a stable idempotency key from data."""
    return hashlib.sha256(data.encode()).hexdigest()[:32]


def is_valid_idempotency_key(key: str) -> bool:
    """Validate idempotency key format."""
    # UUID v4
    uuid_pattern = re.compile(
        r"^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$",
        re.IGNORECASE,
    )
    if uuid_pattern.match(key):
        return True
    
    # Alphanumeric 16-64 chars
    alphanumeric_pattern = re.compile(r"^[a-zA-Z0-9_-]{16,64}$")
    return bool(alphanumeric_pattern.match(key))
